### DISCUSSION TITLE
How to get opengrok indexer logs into syslog

### DISCUSSION THREAD

**Discussion by ohm314** [oracle/opengrok#3840] [Q&A] [answered]
Hi all

Currently I have opengrok setup to write per project log files using the JUL logger, FileHandler and Formatter. This is working OK, but as I'm trying to integrate my opengrok instance into the rest of our development infrastructure, I have to make sure that all logs go to syslog where they can be centralized.

In Python this is all quite easily done since the standard logger library also has a SyslogHandler, but no such luck with JUL (or so it seems?). So, what are my options? Do you have any advice? Can this be done without having to change opengrok? If not, are there any thoughts on how to do this with a patch to opengrok? I am thinking about moving the logging in opengrok to SL4J?

cheers!

**Comment by vladak** [comment, marked as answer]
Firstly, given the log4j stuff that is happening currently (#3837) I'd be very hesitant to switch the logging over from `java.util.logging` (JUL) to something else (even if it was a facade).

I assume you are interested in passing everything from either indexer or the web app to syslog, in particular to a remote syslog server. Not considering the Python based tools at the moment (for simplicity), these contain their own logging (e.g. `opengrok-mirror`).

So, first thing I'd look at is some logging handler that can be added to the list of `handlers` in `logging.properties` that support logging via syslog. https://stackoverflow.com/a/2312463/11582827 suggests to use a JUL to SLF4J bridge (`SLF4JBridgeHandler` - http://www.slf4j.org/legacy.html#jul-to-slf4j) which would then log using http://logback.qos.ch/ configured with http://logback.qos.ch/manual/appenders.html#SyslogAppender that is able to send UDP/TCP Syslog datagrams/segments to remote Syslog server. There seems to be some application performance cost associated with using the ` SLF4JBridgeHandler`.

**Reply by ohm314** [reply]
I expected the log4j debacle would come up and I definitely share your hesitancy!

Exactly, I'm particularly interested in passing everything from the indexer (and possibly the webapp) to syslog. In my use case the local syslog would then forward logs to the centralized infrastructure, but a more generic solution would have to also allow writing to a remote syslog (which should surely happen via the UDP based syslog protocol).
I didn't know about `SLF4JBridgeHandler`. I'll take a look at it as it seems an interesting possibility. It looks there is also a way to mitigate the mentioned performance issues with [LevelChangePropagator](http://logback.qos.ch/manual/configuration.html#LevelChangePropagator).

Thanks!
